// bidibinder/lib/index.js
// Generated by CoffeeScript 1.6.3
(function() {
  var Binding, EQ;

  module.exports = function(opts) {
    return new Binding(opts);
  };

  EQ = function(x, y) {
    return x === y;
  };

  Binding = (function() {
    function Binding(_arg) {
      this.get_a = _arg.get_a, this.get_b = _arg.get_b, this.set_a = _arg.set_a, this.set_b = _arg.set_b, this.equals = _arg.equals;
      if (this.equals == null) {
        this.equals = EQ;
      }
    }

    Binding.prototype._run = function(func) {
      if (this._m !== true) {
        this._m = true;
        func();
        return this._m = false;
      }
    };

    Binding.prototype.touch_a = function() {
      var _this = this;
      return this._run(function() {
        var v;
        if (!_this.equals(_this.get_b(), v = _this.get_a())) {
          return _this.set_b(v);
        }
      });
    };

    Binding.prototype.touch_b = function() {
      var _this = this;
      return this._run(function() {
        var v;
        if (!_this.equals(_this.get_a(), v = _this.get_b())) {
          return _this.set_a(v);
        }
      });
    };

    return Binding;

  })();

}).call(this);
